Implement form elements for the design panel

wassgha opened the pull request:
Addresses #255

<img width="327" alt="Screen Shot 2020-02-09 at 10 48 11 PM" src="https://user-images.githubusercontent.com/591655/74127323-30d6a000-4b8f-11ea-8091-c02edecbdd55.png">

### Changes
- Implements the following form elements for use in the design panel
  - Numeric input
  - Input rows
  - One-state and two-state toggle buttons
  - Spacer
  - Color input (to be used with #180 )
- Simple UI for color presets (non-functional)
- Correctly styles the design panel's headers
- Merges the position, size and rotation panels into one panel
- Uses icons for some actions instead of buttons
- Design update based on new design
- TextStyle panel update
- TextInput
- Unit tests added

### To-do
- [ ] Select input (Will be on separate PR)
- [ ] Panel re-organization (Separate PR for each panel)
- [ ] Slide to change on numeric inputs (In separate PR #389)

/cc @dvoytenko 

See #255

wassgha committed:
Implements generic elements for the design panel

swissspidy commented on assets/src/edit-story/app/uploader/useUploader.js:
```diff
@@ -34,6 +34,9 @@ function useUploader() {
       video: allowedVideoMimeTypes,
+    } = {
+      image: [],
+      video: [],
```
Were there issues without this when running the plugin in a WordPress context?

If it's a storybook thing, then the proposed fix in #157 might make sense.

Otherwise, perhaps the config provider could have some defaults? Is a bit more  DRY.

obetomuniz replied:
@wassgha can you please give me a direction on it?

wassgha replied:
Reverted in a commit. No longer needed.

swissspidy commented on webpack.config.js:
```diff
@@ -39,7 +39,7 @@ const sharedConfig = {
         parallel: true,
         sourceMap: false,
-        cache: true,
+        cache: false,
```
If there's issues with caching we should track that separately. Perhaps there are alternative solutions.

obetomuniz replied:
@wassgha can you please clarify why did you apply this change? Can I revert it or it is a requirement for this PR works properly?

wassgha replied:
Ditto, resolved.

dvoytenko commented on assets/src/edit-story/components/canvas/pagemenu/index.js:
```diff
@@ -29,13 +29,13 @@ import { useCallback } from '@wordpress/element';
-import Add from './add_page.svg';
-import Layout from './layout_helper.svg';
-import Text from './text_helper.svg';
+import Delete from '../../../icons/delete_icon.svg';
```
Hmm. Not 100% sure on this. Some icons I can definitely see as universal. But some icons seem to be so tightly connected with a single use-case, that this would only create more distance between the resource and it's usage. WDYT?

obetomuniz replied:
I will think about it @dvoytenko. I am divided between both approaches as well, but I definitely prefer to manage assets in one place to avoid redundancy and improve performance, anyway, as you mentioned, we lost the local reference for local stuff. In general, what I generally use is aliases like `@assets/images`, do you think that can be useful for this scenario? cc @ndev1991

barklund replied:
I definitely like aliasing in webpack for assets. If we want centralized icons, that would be a good way to do it.

I'd do it in a separate PR though for clarity. Also there's some tooling setup - any opinions on this @swissspidy?

And remember you can even use [emoji aliases for added awesomeness](https://twitter.com/ryanflorence/status/1228749328837009410)! 😉

dvoytenko replied:
As long as we don't start seeing things like:

```
icons/delete.svg
icons/delete_for_this_one_case.svg
icons/delete_for_this_other_case.svg
```

Even this is usually manageable, but IMHO more local placement is cleaner. But I don't have a very strong opinion on this. If there are other benefits from concentrating these resources, I can be persuaded.

swissspidy replied:
Happy to discuss such aliases in an issue, but in my experience they make it more difficult to grasp the code rather than to be actually helpful. But perhaps you can prove me wrong.

dvoytenko replied:
My point was not to force such names. Actually quite the opposite. I was just reflecting that this could happen (and often does). E.g. there naturally might be a couple of very similar delete or duplicate buttons that are used in different situations. Having them co-located along the code clarifies where they are used and how, instead of ad-hoc giving them different names.

dvoytenko commented on assets/src/edit-story/components/form/button.js:
```diff
@@ -0,0 +1,39 @@
+  justify-content: center;
+  align-items: center;
+
+  ${({ expand = true }) => expand && `flex: 1;`}
```
This is not very obvious to me. We don't know if the button is inside a flexbox container or if the default behavior to expand it when it is. I think it's easy enough to customize the styled component's CSS or wrap it at the destination if something like this is necessary.

dvoytenko commented on assets/src/edit-story/components/form/button.js:
```diff
@@ -0,0 +1,39 @@
+import styled from 'styled-components';
+
+const Button = styled.button`
+  color: ${({ theme }) => theme.colors.mg.v3};
```
There's a foreground, but not background color?

obetomuniz replied:
When you say foreground you say the text color @dvoytenko declaration? Just asking to be more in sync with the PR as I and @ndev1991 will take it to work 😄 Also, just to be clear, what you are missing here is the background-color based on theme support, right?

wassgha replied:
To clarify, this is in relation to the fact that it has a constant text color but no background color so it takes the background color of its container meaning that there could be potentially be a case where the text and background are both dark. Since I only expect this to be used for the design panel for now I added a white background color.

dvoytenko commented on assets/src/edit-story/components/form/color.js:
```diff
@@ -0,0 +1,136 @@
+  box-shadow: none !important;
+  outline: none;
+  padding: 0;
+  margin-right: ${({ suffix }) => (Boolean(suffix) ? 12 : 0)}px;
```
Is this a universal `input` or is it meant to be specifically in the design panel? Because there are a lot of assumptions here for a very specific layout. Can these assumptions instead be moved to the container component?

obetomuniz replied:
Just to clarify @dvoytenko: When you say to move this specificity to a container, are you saying to apply specificities only when we reuse the component somewhere as a child component or as a wrapped component? For instance:


```
<Container>
 <Button />
</Container>
```

or

```
const StyledButton = styled(Button)`...` 
```

dvoytenko replied:
Yes. The default component would have no margin:

```
const Input = styled.input`
  ...
`;
```

And the call site, would customize it as needed for a specific situation:

```
import Input from './form/color';

const SpecialInput = styled(Input)`
  margin-right: 8px;
`;
```

dvoytenko commented on assets/src/edit-story/components/form/color.js:
```diff
@@ -0,0 +1,136 @@
+  }
+`;
+
+function Numeric({
```
Rename to `ColorInput`?

wassgha replied:
Renamed. But note to @obetomuniz , the `Input` component inside `Numeric`, `Input` and `Color` could probably be shared so might wanna refactor there.

obetomuniz replied:
Thanks for advance heads up @wassgha

dvoytenko commented on assets/src/edit-story/components/form/color.js:
```diff
@@ -0,0 +1,136 @@
+  margin-left: 5px;
+  margin-right: 5px;
+
+  ${({ expand }) => expand && `flex: 1;`}
```
Ditto.

wassgha committed:
Partially address review comments

ndev1991 committed:
add switch component

swissspidy commented on assets/src/edit-story/components/form/switch.js:
```diff
@@ -0,0 +1,124 @@
+  }
+`;
+
+function Switch({ value, disabled, onChange, onLabel, offLabel }) {
```
Don't we have a `Switch` in `assets/src/edit-story/components/switch/` already? That one seems accessible and could potentially be extended to allow for labels.

ndev1991 replied:
I think that switch is a different one from this slideSwitch. And basically that is a toggle and this is based on switch between two different buttons. So I used two radio buttons instead on checkbox.

ndev1991 replied:
@swissspidy This moved to another PR.  https://github.com/google/web-stories-wp/pull/389

obetomuniz committed:
Address partially PR reviews comments and definitions.

- Remove expand and boxed properties to promote wrap styling of core components
- Refactor input.js and re-use it on color.js and numeric.js
- Remove specific styles to promote wrap styling of core components

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,216 @@
+  padding: 6px 6px;
+  border: 1px solid ${({ theme }) => theme.colors.fg.v3};
+  border-radius: 4px;
+`;
```
👍

barklund commented on assets/src/edit-story/components/form/switch.js:
```diff
@@ -0,0 +1,124 @@
+        onClick={() => onChange(false)}
+        checked={!value}
+        value="off"
+        id="off"
```
ID's have to be unique on the entire page - add a random prefix or similar.

obetomuniz replied:
@ndev1991 in order to move it to another PR, take a look at this improvement. 👍

barklund commented on assets/src/edit-story/components/form/switch.js:
```diff
@@ -0,0 +1,124 @@
+  value,
+  id,
+}))`
+  display: none !important;
```
Is this accessible? Can it still receive focus then if tabbed to or similar and be actived with space bar/enter?

obetomuniz replied:
@ndev1991 in order to move it to another PR, take a look at this improvement. 👍

swissspidy replied:
`display: none` is not accessible, no.

barklund commented on assets/src/edit-story/components/form/switch.js:
```diff
@@ -0,0 +1,124 @@
+        checked={!value}
+        value="off"
+        id="off"
+      />
```
Does it need `aria-labelledby` or is this good enough?

obetomuniz replied:
@ndev1991 in order to move it to another PR, take a look at this improvement. 👍

barklund commented on assets/src/edit-story/components/form/switch.js:
```diff
@@ -0,0 +1,124 @@
+Switch.defaultProps = {
+  disabled: false,
+  onLabel: 'On',
+  offLabel: 'Off',
```
Should be translatable.

obetomuniz replied:
@ndev1991 in order to move it to another PR, take a look at this improvement. 👍

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,216 @@
+          lockRatio &&
+          (newHeight === '' || newWidth === '') &&
+          !(newHeight === '' && newWidth === '')
+        ) {
```
This should be written with some `const`'s above to explain what's going on here. Something like `if (lockRatio && (hasHeightButNotWidth || hasWidthButNotHeight))` or preferably something even better than that 😉

obetomuniz replied:
Cool. I will address this.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,216 @@
+      {/** Width/height & lock ratio */}
+      <Row expand>
+        <BoxedNumeric
+          prefix={__('W', 'web-stories')}
```
This should probably use `_x` too with a note on context - and similarly for `'H'` below.

obetomuniz replied:
Very interesting. It was really uncomfortable to read hehe Nice to know that we have this kind of support. I will replace both. Thanks for the detailed explanation of this @barklund.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,216 @@
+          value={/** TODO */ false}
+          isMultiple={false}
+          onChange={() => {}}
+        />
```
I don't think these are thought to be toggles actually. It's just a button flipping the content, which you can press any number of times without the button changing appearance. I.e. I don't think it has a _"selected"_ state as it's never true or false. It's just a button. I might be wrong, though.

obetomuniz replied:
Agreed. Even considering that we have a toggle UI proposal, it can create confusion, as you said, in the end, only the content is being replaced. I will investigate why it is using the Toggle component instead of the button. Nice catch!

barklund commented on assets/src/edit-story/components/panels/textStyle.js:
```diff
@@ -0,0 +1,115 @@
+      <SelectMenu
+        label={__('Alignment', 'web-stories')}
+        options={alignmentOptions}
+        isMultiple={'' === textAlign}
```
But if the empty string is actually one of the legal values of this dropdown, how do you know if it's just empty or actually the result of different values from a multiselection?

obetomuniz replied:
@ndev1991 in order to move it to another PR, take a look at this improvement. 👍

barklund commented on package.json:
```diff
@@ -25,6 +25,7 @@
     "draft-js-export-html": "^1.4.1",
     "draft-js-import-html": "^1.4.1",
     "draftjs-filters": "^2.3.0",
+    "polished": "^3.4.4",
```
Should it be a `dep` or a `devDep`? @swissspidy?

swissspidy replied:
`dep` makes sense. We should actually move way more there from the dev deps... I'd need to go through the list.

barklund reviewed (commented):
Very clean and nice component library. Great stuff!

obetomuniz committed:
Improve size component on panel based on PR comments.

barklund commented on assets/src/edit-story/components/form/index.js:
```diff
@@ -14,5 +14,14 @@
+export { default as Label } from './label';
+export { default as Row } from './row';
 export { default as SelectMenu } from './select';
+export { default as Spacer } from './row';
```
```suggestion
export { default as Spacer } from './spacer';
```

Commit:
Fix typo

obetomuniz committed:
Merge branch 'master' into design-panel

* master: (52 commits)
  Use clickable icons for range slider in grid view (#345)
  Fix multi-select bug including background element (#367)
  rename completeResize to updateForResizeEvent
  review fixes
  Text resize implementation
  review fixes
  Add missing issue template header (#366)
  Add some issue and PR templates (#304)
  Clean up build folder before each build (#348)
  `restore` should use given array, not recreate (#352)
  Fix some potential PHP warnings and errors (#362)
  fix bad merge
  aria-labelledby for focusable elements
  reduce to flat
  review fixes
  more docs
  switch to mousetrap
  lints
  Prototype keyboard handlers for canvas
  Run static analysis  using  PHPStan on CI (#334)
  ...

ndev1991 committed:
remove switch component from form in order to move it to another pr

swissspidy commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,10 +1,7 @@
----
-
-## Bug Description
+---## Bug Description
```
@obetomuniz Can you undo the changes to the Markdown files within the `.github` folder?

Not 100% sure how that happened, but if it's an issue with the pre-commit hook I'd need to fix that independently.

swissspidy replied:
^ Or @ndev1991

ndev1991 replied:
@swissspidy It seems happened while merging master branch to here. Let me fix this

ndev1991 replied:
@swissspidy Fixed. FYI, empty line added before ## with the pre-commit.

swissspidy replied:
Thanks. I meant all files in `.github/` , not just the one file. Did that now in f17477041c5404476ba5a838261f97a71570aa3c.

Will investigate further why this happened.

ndev1991 committed:
undo bug_report template

swissspidy committed:
Merge branch 'master' into design-panel

swissspidy committed:
Reset .github files back to what’s in master

swissspidy committed:
Remove duplicate dep

barklund commented on assets/src/edit-story/components/form/toggle.js:
```diff
@@ -0,0 +1,107 @@
+          onChange(!value);
+        }
+        evt.stopPropagation();
+      }}
```
This isn't triggered if people click the checkbox directly. And clicking the checkbox directly happens when you tab to it and press <kbd>space</kbd>. So it needs to be an `onChange` on the checkbox rather than `onClick` on the label.

ndev1991 replied:
Refactored to use checkbox onChange event. And also refactored with label container.

barklund commented on assets/src/edit-story/components/form/toggle.js:
```diff
@@ -0,0 +1,107 @@
+Toggle.propTypes = {
+  value: PropTypes.number.isRequired,
+  onChange: PropTypes.func.isRequired,
+  icon: PropTypes.node,
+  uncheckedIcon: PropTypes.node,
+  disabled: PropTypes.bool,
+  boxed: PropTypes.bool,
+  expand: PropTypes.bool,
```
If they aren't required, they should have `defaultProps`. Also, what happens when the icon isn't there?

ndev1991 replied:
DefaultProps added.

barklund commented on assets/src/edit-story/components/form/button.js:
```diff
@@ -0,0 +1,38 @@
+/*
```
I would love to see both unit tests and Storybook stories for these. Nothing complex, just basic toggling/callback checking and a simple story displaying the components in all their states.

swissspidy replied:
Reading my mind here 🙂

obetomuniz replied:
I'm on it. 👍

barklund commented on assets/src/edit-story/components/panels/index.js:
```diff
@@ -107,24 +101,20 @@ export function getPanels(elements) {
             return { type, Panel: FillPanel };
           case BACKGROUND_COLOR:
-            return { type, Panel: BackgroundColorPanel };
+            return { type, Panel: StylePanel };
```
These two lines are pretty weird now. The constant should better reflect which panel it's asking for.

ndev1991 replied:
Done.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+  opacity: 0.54;
+`;
+
 function SizePanel({ selectedElements, onSetProperties }) {
```
Since this is now a lot more than a size panel, it should be renamed and related properties too (like `PanelTypes.SIZE`)

obetomuniz replied:
@barklund Just to clarify how to attack this, as far as I understand, this panel is grouped Sizing and Positioning controls, right? So, can you pls confirm if the list below are the expected changes? Thanks.

- Rename the file from `size.js` to `sizePosition.js`
- Rename `SizePanel` component name to `SizePositionPanel`
- Update references that are using `SIZE_POSITION` instead of `SIZE` (e.g. `PanelTypes.SIZE_POSITION`)

barklund replied:
Sounds good. And make sure all references to old panels are removed.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+        const hasHeightButNotWidth = newHeight === '' || newWidth === '';
+        const hasWidthButNotHeight = newHeight === '' && newWidth === '';
+
+        if (lockRatio && hasHeightButNotWidth && !hasWidthButNotHeight) {
```
This makes no sense now. The variable naming has nothing to do with their contents. My idea here was of course to actually rewrite the logic to reflect the variable naming:

```js
        const hasHeightButNotWidth = newHeight !== '' && newWidth === '';
        const hasWidthButNotHeight = newWidth !== '' && newHeight === '';

        if (lockRatio && (hasHeightButNotWidth || hasWidthButNotHeight)) {
```

Of course triple-check that this has the same logical meaning as the previous.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+
+        if (lockRatio && hasHeightButNotWidth && !hasWidthButNotHeight) {
+          const ratio = oldWidth / oldHeight;
+          if (newWidth === '') {
```
You can then reuse the above variable here:

```js
          if (hasHeightButNotWidth) {
```

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+          const ratio = oldWidth / oldHeight;
+          if (newWidth === '') {
+            update.width = dataPixels(newHeight * ratio);
+          } else if (newHeight === '') {
```
And the condition here would not be needed, as the implication from all the above would be, that `hasWidthButNotHeight` has to be `true` here.

ndev1991 replied:
We don't need else if. Just else needed.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+    (evt) => {
+      onSetProperties(({ width: oldWidth, height: oldHeight }) => {
+        const { height: newHeight, width: newWidth, ...rest } = state;
+        const update = { width: newWidth, height: newHeight, ...rest };
```
Isn't this identical to just `const update = [...state];`? And then you don't need `rest` at all?

ndev1991 replied:
Done.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+              ...state,
+              x: isNaN(value) || value === '' ? '' : parseFloat(value),
+            })
+          }
```
This handler is used many times - maybe create a generic handler above as:

```js
const handleNumberChange = useCallback((property) => (value) => setState((state) => ({
   ...state,
   [property]: isNaN(value) || value === '' ? '' : parseFloat(value),
}))
```

And then where used simply do:

```js
onChange={handleNumberChange('x')}
```

ndev1991 replied:
Done.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+          disabled={isFill}
+        />
+        <FlipButton>
+          <FlipButtonIcon as={FlipHorizontal} />
```
What? Can you use `as` for this? That's pretty cool! TIL

ndev1991 replied:
I checked this part and it's not working as expected. So I updated this parts.

barklund commented on assets/src/edit-story/components/panels/size.js:
```diff
@@ -18,101 +18,218 @@
+          }}
+          disabled={isFill}
+        />
+        <FlipButton>
```
`/** TODO? */`

barklund commented on assets/src/edit-story/theme.js:
```diff
@@ -47,6 +47,8 @@ const theme = {
     mg: {
       v1: '#616877',
       v2: '#DADADA',
+      v3: '#3C4043',
```
Note that this color is not actually used as is. The text in question has an opacity of 0.86 resulting in a color almost identical to `bg.v5`. So I think you can safely replace all instances of `mg.v3` with that and remove it from here.

wassgha replied:
Let's use the real color here (not `bg.v5`, it's subtle but it really does make a difference) however let's try to play with opacity and deduce the color from the existing colors we have and `polished`'s `rgba`.

ndev1991 replied:
Done

barklund commented on assets/src/edit-story/theme.js:
```diff
@@ -47,6 +47,8 @@ const theme = {
       v1: '#616877',
       v2: '#DADADA',
+      v3: '#3C4043',
+      v4: '#80868B',
```
Similarly this color is only ever used in combination with an opacity of .55 resulting in `#b9bcbf`, which is almost identical to `fg.v4` at `#b3b3b3`. This variable here can be removed.

wassgha replied:
ditto here, let's still use the exact color but try to get it from the existing colors and remove these two added colors

ndev1991 replied:
Use the exact color and opacity from figma and used rgba from polished.

barklund reviewed (changes requested):
A lot of smaller comments, but also some bigger ones about (visual and unit) testing - though it's actually pretty simple to do.

swissspidy committed:
Merge branch 'master' into design-panel

swissspidy committed:
Merge branch 'master' into design-panel

ndev1991 committed:
lint error

ndev1991 committed:
lint error

obetomuniz committed:
Merge branch 'master' into design-panel

* master:
  Set up code owners (#379)
  Open the preview in the same tab (#420)
  Implement correct interactive library tabs (#361)
  New page background panel (#383)
  Partially implement the elements panel (#266)

obetomuniz committed:
Add improvements based in PR reviews comments.

ndev1991 committed:
design panel design update #1

obetomuniz committed:
Add unit tests

obetomuniz commented on assets/src/edit-story/components/panels/videoPoster.js:
```diff
@@ -40,6 +40,7 @@ const Button = styled.button`
+  object-fit: contain;
 `;
 
 function VideoPosterPanel({ selectedElements, onSetProperties }) {
```
Adding unit tests for this component I figured issues related to WordPress. It is reporting `ReferenceError: wp is not defined`, its a problem related to the `useMediaPicker` hook which is using `wp.media` directly and Jest can't recognize this, so I would like to get additional insights or suggestions from the team. Do you already figure out a problem like this? cc @barklund @ndev1991 @swissspidy

swissspidy replied:
You should be able to mock anything under the `wp` global by defining it in `tests/js/setup-globals.js`, i.e. `global.wp = jest.fn()`. In the old plugin we used something like this for similar situations:

```js
// Set up `wp.*` aliases.
global.wp = {
	media: {
		controller: {
			Library: jest.fn(),
			Cropper: {
				extend: jest.fn(),
			},
		},
		View: {
			extend: jest.fn(),
		},
		view: {
			Toolbar: {
				Select: {
					extend: jest.fn(),
				},
			},
			MediaFrame: {
				Select: {
					extend: jest.fn(),
				},
			},
		},
	},
};
```

You could also mock globals just in the specific test file you are writing instead of globally this way.

obetomuniz replied:
Nice! Thank you for the suggestion @swissspidy. I will take a look on this approach 👍

swissspidy replied:
We should also look into how we can abstract this usage better to simplify storybook integration.

cc @spacedmonkey

ndev1991 committed:
design panel style update and component update #2

swissspidy commented on assets/src/edit-story/components/panels/test/textStyle.js:
```diff
@@ -0,0 +1,67 @@
+import TextStyle from '../textStyle';
+
+function arrange(children = null) {
+  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
```
Need to use `FontProvider` here as well to fix test failures

swissspidy commented on assets/src/edit-story/components/panels/textStyle.js:
```diff
@@ -0,0 +1,287 @@
+      <Row>
+        {fonts && (
+          <SelectMenu
+            ariaLabel={__('Font family', 'web-stories')}
```
I see `ariaLabel` being used a lot, but none of the components actually support it.

For example, `SelectMenu` does not have any `ariaLabel` prop defined.

ndev1991 replied:
Good points! aria-label added to input and selectmenu

swissspidy commented on assets/src/edit-story/components/form/select.js:
```diff
@@ -43,10 +70,12 @@ function SelectMenu({
-    <Group disabled={disabled}>
-      <Label>{label}</Label>
+    <Container disabled={disabled} flexBasis={flexBasis} flexGrow={flexGrow}>
+      {label && <Label>{label}</Label>}
```
If `label` is optional, the prop types definition should be adjusted accordingly (currently marked as `isRequired`.

ndev1991 replied:
PropTypes adjusted

swissspidy commented on assets/src/edit-story/components/panels/textStyle.js:
```diff
@@ -0,0 +1,287 @@
+        />
+      </Row>
+      {/* TODO: Update padding logic */}
+      {/* <InputGroup
```
Remove this commented out code? Or what's its purpose here?

ndev1991 replied:
There was original logic for padding adjustment. Removed and moved TODO to upper level.

swissspidy commented on assets/src/edit-story/components/panels/textStyle.js:
```diff
@@ -0,0 +1,287 @@
+      <Row>
+        <Label>{__('Padding', 'web-stories')}</Label>
+        <BoxedNumeric
+          suffix={_x('H', 'The Horizonatal padding', 'web-stories')}
```
```suggestion
          suffix={_x('H', 'The Horizontal padding', 'web-stories')}
```

ndev1991 replied:
Thank you for your point @swissspidy . Fixed typo

ndev1991 committed:
update various issues based on comments

ndev1991 committed:
add apiProvider to fontProvider working

ndev1991 committed:
fix color on form button based on design

obetomuniz committed:
Add initial videoPost panel tests

obetomuniz committed:
Merge branch 'design-panel' of github.com:google/web-stories-wp into design-panel

* 'design-panel' of github.com:google/web-stories-wp:
  fix color on form button based on design

obetomuniz committed:
Update unit tests to cover recent changes based on new designs

swissspidy commented on package.json:
```diff
@@ -101,7 +101,8 @@
     "webpack-cli": "^3.3.11",
-    "webpackbar": "^4.0.0"
+    "webpackbar": "^4.0.0",
+    "whatwg-fetch": "^3.0.0"
```
What about using something like jest-fetch-mock instead?

obetomuniz replied:
In general, I use `whatwg-fetch` to mock `window.fetch`, but I don't have problems to migrate to `jest-fetch-mock` 😃  I'm on it. Thanks for the suggestion @swissspidy

miina commented on assets/src/edit-story/components/panels/colorPreset.js:
```diff
@@ -14,22 +14,59 @@
 function ColorPresetPanel() {
   return (
     <Panel name="colorpreset">
       <PanelTitle isPrimary>{__('Color presets', 'web-stories')}</PanelTitle>
```
This seems to be missing a `+` for adding a new preset.

miina commented on assets/src/edit-story/components/form/button.js:
```diff
@@ -0,0 +1,39 @@
+
+const Button = styled.button`
+  background: ${({ theme }) => theme.colors.fg.v1};
+  color: ${({ theme }) => rgba(theme.colors.fg.v1, 0.1)};
```
The background and color seem to have the same color here -- probably the text is not really visible like this? :)

miina commented on assets/src/edit-story/components/panels/font.js:
```diff
@@ -106,39 +129,79 @@ function FontPanel({ selectedElements, onSetProperties }) {
-              fontFallback: currentFontFallback,
-            });
-          }}
+          ariaLabel={__('Font style', 'web-stories')}
```
Note: This select is most likely for `fontWeight` as the removed code above -- `fontStyle` is resolved via a Toggle:
<img width="378" alt="Screenshot 2020-03-03 at 09 02 08" src="https://user-images.githubusercontent.com/3294597/75754820-ae747280-5d2d-11ea-9a05-6101b60a1ba4.png">

ndev1991 replied:
Panel added to textStyle panel. I think font panel is not necessary since we have a textStyle panel.

miina commented on assets/src/edit-story/components/panels/font.js:
```diff
@@ -106,39 +129,79 @@ function FontPanel({ selectedElements, onSetProperties }) {
+          suffix={<VerticalOffset />}
+          isMultiple={fontSize === ''}
+          onChange={(value) =>
+            setState({ ...state, fontSize: parseInt(value) })
```
This is probably for `lineHeight` (which should already be functional, too).

miina commented on assets/src/edit-story/components/panels/font.js:
```diff
@@ -106,39 +129,79 @@ function FontPanel({ selectedElements, onSetProperties }) {
+          suffix={<HorizontalOffset />}
+          isMultiple={fontSize === ''}
+          onChange={(value) =>
+            setState({ ...state, fontSize: parseInt(value) })
```
This is probably for `letterSpacing` (which should already be functional, too), in another panel though (the `textStyle` panel).

I'm happy to address the functionality of this in the Design Panel functionality PRs instead though.

miina commented on assets/src/edit-story/components/form/numeric.js:
```diff
@@ -0,0 +1,126 @@
+        prefix={prefix}
+        suffix={suffix}
+        label={label}
+        value={`${value}${symbol}`}
```
Is it intended that the value would include the `symbol` on submitting?

miina commented on assets/src/edit-story/components/panels/test/actions.js:
```diff
@@ -0,0 +1,62 @@
+/*
```
The Actions panel was really a temporary panel until removing an element via keyboard was implemented (which should be done now). Both the action panel and the tests for it should be obsolete.

(Can be removed elsewhere out of this PR though)

miina commented on assets/src/edit-story/components/panels/textStyle.js:
```diff
@@ -0,0 +1,281 @@
+/*
```
This seems to be largely very similar to the fonts panel, same comments here as for the font's panel.

obetomuniz replied:
Indeed. Both are very similar @miina Can you please confirm if we still need that `font` panel? Font's Panel looks like not being in use and redundant.

miina replied:
A bit delayed answer: That's correct, the font's panel is redundant, it will be removed within #412

obetomuniz replied:
Sure! Let me reference this on #255. Thanks for sync up @miina

miina reviewed (commented):
Note that there are parallel PR-s for the functionality of the Design Panel parts which also do shuffling around the panels and adding missing functionality.

As long as the form elements all exist and are designed I'm happy to do all the refactoring and functionality fixes for the things that broke, in those PRs to get this merged faster and for having more smaller follow-up PRs instead. /cc @swissspidy

ndev1991 committed:
font style panel

ndev1991 committed:
merge

miina commented on assets/src/edit-story/components/panels/colorPreset.js:
```diff
@@ -14,22 +14,59 @@
+/**
+ * External dependencies
+ */
+import styled from 'styled-components';
```
External dependencies are defined twice, there's one above already.

ndev1991 committed:
remove duplicated external dependencies section

obetomuniz committed:
Replace whatwg-fetch to promote jest-fetch-mock while testing and mocking Fetch API

obetomuniz committed:
Merge branch 'design-panel' of github.com:google/web-stories-wp into design-panel

* 'design-panel' of github.com:google/web-stories-wp:
  remove duplicated external dependencies section

swissspidy commented on .eslintrc:
```diff
@@ -8,6 +8,12 @@
     "prettier/react"
   ],
   "plugins": ["jsx-a11y", "header"],
+  "env": {
```
Can you clarify what exactly each change here does and why this was suddenly necessary now?

Because some of these envs are already set, others only make sense in a specific context.

obetomuniz replied:
Yeah, I see. It was necessary because of a `no-undef` ESLint Error while using `fetch` globally via `jest-fetch-mock`. It includes the environments that expose globals when it is defined (e.g. `global.fetch` and `window.fetch`). What do you think?

obetomuniz replied:
I'll be using `global.fetch` instead 🥂

swissspidy replied:
Works for me! Thanks!

obetomuniz committed:
Address some PR review comments

obetomuniz committed:
Remove unnecessary global config statement.
